[WGSL] Convert AST into string representation

djg opened the pull request:
#### 7e29ffc25c726581465a62874ef92ff8ea7bfabc
<pre>
[WGSL] Convert AST into string representation
<a href="https://bugs.webkit.org/show_bug.cgi?id=246791">https://bugs.webkit.org/show_bug.cgi?id=246791</a>
rdar://problem/101366942

Reviewed by Myles C. Maxfield.

As a debugging aid, AST::Dumper converts a given WGSL AST into a string
representation that approximates the original WGSL source code.

Tested by ASTDumperTests.

* Source/WebGPU/WGSL/AST/ASTDumper.cpp: Added.
(WGSL::AST::Indent::Indent):
(WGSL::AST::bumpIndent):
(WGSL::AST::Dumper::visitVector):
(WGSL::AST::Dumper::toString):
(WGSL::AST::Dumper::visit):
(WGSL::AST::dumpNode):
(WGSL::AST::dumpAST):
* Source/WebGPU/WGSL/AST/ASTDumper.h: Added.
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ASTDumperTests.cpp: Added.
(TestWGSLAPI::toString):
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(logCompilationError): Deleted.
* Tools/TestWebKitAPI/Tests/WGSL/TestWGSLAPI.cpp: Added.
(TestWGSLAPI::logCompilationError):
* Tools/TestWebKitAPI/Tests/WGSL/TestWGSLAPI.h: Added.

Canonical link: <a href="https://commits.webkit.org/255812@main">https://commits.webkit.org/255812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5e873c8da5e7e2286102d4931187886c52d10e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2844 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103300 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163623 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2855 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31122 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86001 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99311 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2033 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80093 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29061 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83957 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37507 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35349 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4013 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39225 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41160 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/AST/ASTDumper.cpp:
```diff
@@ -0,0 +1,403 @@
+    SetForScope<String> m_scope;
+};
+
+static auto bumpIndent(Dumper& dumper) -> Indent
```
why not `static Indent bumpIndent(Dumper& dumper)`?

litherum commented on Source/WebGPU/WGSL/AST/ASTDumper.cpp:
```diff
@@ -0,0 +1,403 @@
+#define NEWLINE()   m_out.print("\n");
+#define NL_INDENT() m_out.print("\n", m_indent)
+#define NODENT(...) m_out.print(__VA_ARGS__)
+#define SPACE()     m_out.print(" ");
```
I think we should not use `#define`s for these

djg replied:
So you'd rather see `m_out.print(m_indent, "enable ", directive.name(), ";");` instead of `INDENT("enable ", directive.name(), ";");`?

litherum commented on Source/WebGPU/WGSL/AST/ASTDumper.h:
```diff
@@ -0,0 +1,101 @@
+
+namespace WGSL::AST {
+
+class Dumper final : public Visitor {
```
Can we call it "StringDumper" or something? Just "Dumper" seems a bit too vague

litherum commented on Source/WebGPU/WGSL/AST/ASTVisitor.h:
```diff
@@ -37,53 +37,53 @@ class Visitor {
     virtual ~Visitor() = default;
 
     // Shader Module
-    virtual void visit(AST::ShaderModule&);
-    virtual void visit(AST::GlobalDirective&);
```
Normally I'd recommend doing this stuff in a different patch, but I guess it's fine here since we're trying to get off the ground.

litherum commented on Tools/TestWebKitAPI/Tests/WGSL/ASTDumperTests.cpp:
```diff
@@ -0,0 +1,71 @@
+        "    var pos = array<Vec2<f32>, 3>(Vec2<f32>(0.000000, 0.500000), Vec2<f32>(-0.500000, -0.500000), Vec2<f32>(0.500000, -0.500000));\n"
+        "    return Vec4<f32>(pos[VertexIndex], 0.000000, 1.000000);\n"
+        "}\n\n\n"_str);
+}
```
This is may be heresy, but do we actually need to test the dumper? It's not web-exposed, and just for debug development, right?

djg replied:
All code needs tests, right?

litherum commented on Source/WebGPU/WGSL/AST/ASTDumper.cpp:
```diff
@@ -0,0 +1,403 @@
+    visit(expression.target());
+    NODENT("(");
+    if (!expression.arguments().isEmpty()) {
+        auto indent = bumpIndent(*this);
```
I'm not looking too deeply at the guts of this, because IIUC it's just for local debugging

djg committed:
[WGSL] Convert AST into string representation
https://bugs.webkit.org/show_bug.cgi?id=246791
rdar://problem/101366942

Reviewed by Myles C. Maxfield.

As a debugging aid, AST::Dumper converts a given WGSL AST into a string
representation that approximates the original WGSL source code.

Tested by ASTDumperTests.

* Source/WebGPU/WGSL/AST/ASTDumper.cpp: Added.
(WGSL::AST::Indent::Indent):
(WGSL::AST::bumpIndent):
(WGSL::AST::Dumper::visitVector):
(WGSL::AST::Dumper::toString):
(WGSL::AST::Dumper::visit):
(WGSL::AST::dumpNode):
(WGSL::AST::dumpAST):
* Source/WebGPU/WGSL/AST/ASTDumper.h: Added.
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ASTDumperTests.cpp: Added.
(TestWGSLAPI::toString):
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(logCompilationError): Deleted.
* Tools/TestWebKitAPI/Tests/WGSL/TestWGSLAPI.cpp: Added.
(TestWGSLAPI::logCompilationError):
* Tools/TestWebKitAPI/Tests/WGSL/TestWGSLAPI.h: Added.

Canonical link: https://commits.webkit.org/255812@main
